static/js/message_list_view.js

```diff
@@ -1445,4 +1445,76 @@ export class MessageListView {
             message_container.status_message = false;
         }
     }
+
+    update_sticky_recipient_headers() {
+        const rows_length = this._rows.size;
+        if (!rows_length) {
+            /* No headers are present */
+            return;
+        }
+        /* Intentionally remove sticky headers class here to make calculations simpler. */
+        $(".sticky_header").removeClass("sticky_header");
+        /* visible_top is navbar top position + height for us. */
+        const visible_top = $message_viewport.message_viewport_info().visible_top;
+        /* We need date to be properly visible on the header, so partially visible headers
+        who are about to be scrolled out of view are not acceptable. */
+        const $table = rows.get_table(this.table_name);
+        const $headers = $table.find(".message_header");
+        const header_height = $headers[0].getBoundingClientRect().height;
+
+        /* There are 3 types elements we expect to be present at this pixel:
+            * message header
+            * Date separator
+            * Z logo at the top of message feed
+
+            If we encounter either loading indicator or date separator, we set sticky header to the next recipient row. */
+        const elements_below_navbar = document.elementsFromPoint(
+            window.innerWidth / 2,
+            visible_top + header_height - 1,
+        );
+        let $sticky_header = $(
+            elements_below_navbar.filter((element) => element.classList.contains("message_header")),
+        );
+        const message_row = elements_below_navbar.filter((element) =>
+            element.classList.contains("message_row"),
+        );
+        const date_row = elements_below_navbar.filter((element) =>
+            element.classList.contains("date_row"),
+        );
+
+        /* Set correct date for the sticky_header. */
+        let $message_row;
+        if (date_row.length) {
+            /* There is always a message header present next to a date_row. */
+            const recipient_row = date_row[0].nextElementSibling;
+            $sticky_header = $(recipient_row).find(".message_header");
+            $message_row = $sticky_header.next();
+        } else if (!$sticky_header.length) {
+            /* If user is at the top of the scroll container, there is no sticky
+               header, so we just set date of the header to the first visible message. */
+            $sticky_header = $headers.first();
```

Comment:
This reads oddly given we're about to set the `$sticky_header` variable ... -- maybe just "If the user is at the top of the scroll container, the header is for the first message group, and we can display the date for the first visible message."

Reply:
Sure, done!